r/tests/testthat/test-dplyr-mutate.R

```diff
@@ -344,20 +344,21 @@ test_that("print a mutated table", {
       select(int) %>%
       mutate(twice = int * 2) %>%
       print(),
-'Table (query)
+'InMemoryDataset (query)
 int: int32
 twice: expr
 
 See $.data for the source Arrow object',
   fixed = TRUE)
 
   # Handling non-expressions/edge cases
+  skip("InMemoryDataset$Project() doesn't accept array (or could it?)")
   expect_output(
     Table$create(tbl) %>%
       select(int) %>%
       mutate(again = 1:10) %>%
```

Comment:
This isn't supported because (for most datasets) we don't know the number of rows before scanning, so it's unclear if augmenting the table with an explicit column is valid. It's *possible* to add support for projections like this to Expression but it'd mean they were no longer strictly scalar (since they'd only be valid in the case where the result of the scan happened to have the expected number of rows). This would complicate scanning drastically, since we'd have to accommodate array literals in each scanned batch

I'd expect that augmenting with an explicit array is only done with small, cheap scans? If so, one way to work around this would be to auto-`collect()` the preceding steps and handle the explicit array with `Table$AddColumn`. In any case it might be worthwhile to maintain an `is_scan_of_in_memory_data` flag of the dplyr query as a hint of whether auto `collect()`ion is reasonable

@westonpace what do you think?

Reply:
Are you proposing that I do this detection in R or that it happen in C++?

Reply:
None of the examples on https://dplyr.tidyverse.org/reference/mutate.html actually use this form and I have a hard time understanding why someone might want to do this?

Furthermore, this question https://stackoverflow.com/questions/60582562/another-length-error-using-dplyr-mutate-and-if-else shows some of the confusion you run into with something like this.

From an SQL perspective the proper way to add in a new column would be to join.  This is sort of a "join without a common key" which raises a few eyebrows in this question: https://stackoverflow.com/questions/1198124/combine-two-tables-that-have-no-common-fields

Also, would the vector be the same length as a single batch?  Or the entire table?  If it's the entire table then it's going to force the table to be processed in order which is undesirable as well.

I think I'd want to see a valid use case before investing effort.

Reply:
The workaround I proposed and detection of an in memory scan would occur in R. If we don't *need* to support mutate with vectors then I agree that it'd be ideal to avoid dealing with it